### PR TITLE
projected compression for llmrandom checkpoints

### PR DESCRIPTION
**Pull request by crewtool**

### CODE REVIEW

**Comment by madragonse** [comment]
Have functional requirements been tested? eg. loading weights from llmr checkpoint into nano fw and eveluating by running eval tests or training? @crewtool

File: src/projected_compression/compression.py [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+
+        cloned_data = block.ff_layer.norm.weight.data.clone()
+        block.ff_layer.norm.weight = torch.nn.Parameter(cloned_data[dmodel_top_indices])
+        block.ff_layer.norm.normalized_shape = tuple(block.ff_layer.norm.weight.shape)
+
+
+def init_compression(model: nn.Module, dmodel, dff):
+    # Freeze all parameters
+    for param in model.parameters():
+        param.requires_grad = False
```
Comment (madragonse) [inline src/projected_compression/compression.py:123]: Where projection weights are unfrozen if all parameters are being frozen here? I can't find it in the init_projections method if it is there.

File: src/core/model.py [hunk 2]
```diff
@@ -15,7 +15,7 @@
     LayerNorm as LayerNorm,
```
Comment (madragonse) [inline src/core/model.py:15]: Is it used in PC project, any of code from this file? I have seed copies of this model.py classes in projected_compression/model.py.

File: src/projected_compression/model.py [hunk 3]
```diff
@@ -0,0 +1,540 @@
+from functools import partial
```
Comment (madragonse) [inline src/projected_compression/model.py:1]: As above - is it only model.py currently used in PC experiments, is the transition compreted, or not and it shouldn't be, what is the concept fot code struct?

File: configs/llm-random.yaml [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+    dataset_type: c4
+    num_workers: 0 # Change me!
+    seed: 123
+    sequence_length: 32
+    shuffle: true
+    total_batch_size: 32
+    use_new_sampling_method: true
+    world_size_independent: false
+
+  distributed: null
```
[the remark this answers is not in the document]
Comment (crewtool) [reply configs/llm-random.yaml:174]: no

File: src/core/conversion_from_llmrandom.py [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+
+def remap_llmrandom_state_dict_to_nano(llmrandom_dict):
+
+    replacement_mappings = [
+        # Embedding
+        (r"embedding_layer\.layers\.0\.weight", "embedding.embedding.embedding.weight"),
+        # Attention projections
+        (
+            r"encoder\.blocks\.block_(\d+)\.block\.residual_attention\.layer\.pre_norm\.weight",
+            r"encoder.blocks.\1.attention_layer.norm.weight",
```
[the remark this answers is not in the document]
Comment (crewtool) [reply src/core/conversion_from_llmrandom.py:14]: its correct

File: src/core/conversion_from_llmrandom.py [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+        ),
+        (
+            r"encoder\.blocks\.block_(\d+)\.block\.residual_attention\.layer\.attention\.input_projection_q\.weight",
+            r"encoder.blocks.\1.attention_layer.layer.q_proj.weight",
+        ),
+        (
+            r"encoder\.blocks\.block_(\d+)\.block\.residual_attention\.layer\.attention\.input_projection_k\.weight",
```
[the remark this answers is not in the document]
Comment (crewtool) [reply src/core/conversion_from_llmrandom.py:21]: yes

File: src/core/conversion_from_llmrandom.py [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+            f"encoder.blocks.{n_layer}.attention_layer.layer.q_proj.weight"
+        ] = q.reshape(heads * dhead, -1)
+        remapped_state_dict[
+            f"encoder.blocks.{n_layer}.attention_layer.layer.k_proj.weight"
+        ] = k.reshape(heads * dhead, -1)
+        remapped_state_dict[
+            f"encoder.blocks.{n_layer}.attention_layer.layer.v_proj.weight"
+        ] = v.reshape(heads * dhead, -1)
+
+def load_llmrandom_checkpoint(load_config, model):
```
[the remark this answers is not in the document]
Comment (crewtool) [reply src/core/conversion_from_llmrandom.py:108]: We do not have llama oprtimizer state, theregore no
